File: packages/gateway/src/cartello.ts
/**
 * **Il cartello**: dove il computer scrive come trovarlo, quando il suo
 * indirizzo da fuori cambia nome.
 *
 * Chiesto l'11 settembre 2026, dopo una sera passata fuori casa: «la
 * connessione e' la cosa piu' importante, deve sempre essere affidabile e
 * riconnettere bene gli utenti... aspetto il tuo fix DEFINITIVO».
 *
 * ## Il buco che chiude
 *
 * Da fuori si passa dal tunnel gratuito di Cloudflare, e il suo nome cambia
 * quando il tunnel riparte. Il telefono prova gli indirizzi che ha in tasca, e
 * se sono tutti morti non ha modo di sapere quello nuovo: per saperlo dovrebbe
 * parlare col computer, e per parlarci gli serve l'indirizzo. La 1.3.1 ha curato
 * il rientro a mano — un messaggio da mandare su WhatsApp — e questo lo fa da
 * solo, senza che nessuno debba accorgersene.
 *
 * ## Come funziona
 *
 * Il computer scrive i suoi indirizzi da fuori su una bacheca pubblica —
 * ntfy.sh, gratuita e senza account — sotto un nome diverso per ogni telefono,
 * ricavato dalla chiave di quel telefono. Il telefono, quando non trova il
 * computer, legge il suo cartello, controlla la firma, e bussa all'indirizzo
 * nuovo con la solita porta `/io`.
 *
 * ## Perche' ci si puo' fidare di una bacheca pubblica
 *
 * - **Il nome del cartello si ricava dalla chiave del telefono**: chi non ha la
 *   chiave non sa dove guardare. Sono 128 bit, non si indovinano.
 * - **Quello che c'e' scritto e' firmato** con la stessa chiave: chi indovinasse
 *   il nome non potrebbe scriverci un indirizzo suo, perche' il telefono lo
 *   butterebbe via.
 * - **E comunque il telefono bussa con `/io`**: se dall'altra parte non c'e' il
 *   suo computer, non entra niente.
 *
 * Sulla bacheca finisce l'indirizzo del tunnel e basta — lo stesso che e' gia'
 * pubblico su Internet, e che senza chiave risponde 401 a tutto. Non ci finisce
 * la chiave, non ci finisce la rete di casa.
 *
 * ⚠ **La stessa ricetta sta nell'app**, in `Cartello.kt`. Due lingue e una
 * ricetta sola: `prova-cartello.mjs` e `CartelloTest.kt` calcolano lo stesso
 * esempio e aspettano gli stessi numeri. Se una delle due cambia, cade la sua
 * prova — che e' l'unico modo di tenere allineate due cose scritte in due
 * lingue.
 */

import { createHash, createHmac, timingSafeEqual } from "node:crypto";

/** La bacheca. Gratuita, senza account, e tiene i messaggi dodici ore. */
export const BACHECA = "https://ntfy.sh";

/** La forma del cartello. Sale se un giorno cambia cosa c'e' scritto. */
const VERSIONE = 1;

/** Il nome del cartello di un telefono: si ricava dalla sua chiave, e non la rivela. */
export function argomentoDi(token: string): string {
  const impronta = createHash("sha256").update("daprod-cartello-nome:" + token).digest("hex");
  return "daprod_" + impronta.slice(0, 32);
}

/** La chiave della firma: anche questa dalla chiave del telefono, e diversa dal nome. */
function chiaveDi(token: string): Buffer {
  return createHash("sha256").update("daprod-cartello-chiave:" + token).digest();
}

/** Quello che si firma: una riga per pezzo, sempre nello stesso ordine. */
function testoDaFirmare(pc: string, quando: number, basi: string[]): string {
  return ["daprod-cartello/" + VERSIONE, pc, String(quando), ...basi].join("\n");
}

export function firmaDi(token: string, pc: string, quando: number, basi: string[]): string {
  return createHmac("sha256", chiaveDi(token)).update(testoDaFirmare(pc, quando, basi)).digest("hex");
}

/** Quello che sta scritto sul cartello. */
export interface Cartello {
  v: number;
  /** Il computer che l'ha scritto: l'id con cui si annuncia sulla rete. */
  pc: string;
  quando: number;
  /** Gli indirizzi da fuori, dal piu' vicino. */
  basi: string[];
  firma: string;
}

export function scriviCartello(
  token: string,
  pc: string,
  basi: string[],
  quando: number = Date.now(),
): Cartello {
  return { v: VERSIONE, pc, quando, basi, firma: firmaDi(token, pc, quando, basi) };
}

/**
 * Legge un cartello e lo tiene **solo se la firma torna**. Serve alle prove e a
 * chi un giorno volesse leggerlo dal computer: il telefono fa lo stesso conto
 * in `Cartello.kt`.
 */
export function leggiCartello(testo: string, token: string): Cartello | null {
  try {
    const c = JSON.parse(testo) as Partial<Cartello>;
    if (c.v !== VERSIONE || typeof c.pc !== "string" || typeof c.quando !== "number") return null;
    if (!Array.isArray(c.basi) || typeof c.firma !== "string") return null;
    const basi = c.basi.filter((b): b is string => typeof b === "string");
    const attesa = Buffer.from(firmaDi(token, c.pc, c.quando, basi), "hex");
    const arrivata = Buffer.from(c.firma, "hex");
    if (arrivata.length !== attesa.length || !timingSafeEqual(arrivata, attesa)) return null;
    return { v: VERSIONE, pc: c.pc, quando: c.quando, basi, firma: c.firma };
  } catch {
    return null;
  }
}

export interface EsitoCartelli {
  /** Quanti cartelli sono stati scritti. */
  scritti: number;
  /** Cosa e' andato storto, uno per cartello non scritto. */
  falliti: string[];
}

/**
 * Scrive il cartello di ogni telefono. Non solleva mai: una bacheca che non
 * risponde e' un cartello in meno, non una suite che si ferma.
 *
 * `manda` e `bacheca` si passano da fuori per le prove, che non vanno su
 * Internet.
 */
export async function pubblicaCartelli(
  tokens: string[],
  pc: string,
  basi: string[],
  opzioni: { bacheca?: string; manda?: typeof fetch; quando?: number } = {},
): Promise<EsitoCartelli> {
  const bacheca = (opzioni.bacheca ?? BACHECA).replace(/\/+$/, "");
  const manda = opzioni.manda ?? fetch;
  const quando = opzioni.quando ?? Date.now();
  const esito: EsitoCartelli = { scritti: 0, falliti: [] };
  for (const token of [...new Set(tokens.filter((t) => t))]) {
    const argomento = argomentoDi(token);
    try {
      const risposta = await manda(`${bacheca}/${argomento}`, {
        method: "POST",
        body: JSON.stringify(scriviCartello(token, pc, basi, quando)),
        signal: AbortSignal.timeout(10_000),
      });
      if (risposta.ok) esito.scritti += 1;
      else esito.falliti.push(`${argomento.slice(0, 14)}…: risponde ${risposta.status}`);
    } catch (err) {
      esito.falliti.push(`${argomento.slice(0, 14)}…: ${err instanceof Error ? err.message : String(err)}`);
    }
  }
  return esito;
}

File: apps/mobile/app/src/main/java/it/daprod/suite/net/Cartello.kt
package it.daprod.suite.net

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.security.MessageDigest
import java.util.concurrent.TimeUnit
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * **Il cartello**: dove il computer scrive come trovarlo, quando il suo
 * indirizzo da fuori cambia nome.
 *
 * Chiesto l'11 settembre 2026, dopo una sera fuori casa: «la connessione e' la
 * cosa piu' importante, deve sempre essere affidabile e riconnettere bene gli
 * utenti».
 *
 * ## Il buco che chiude
 *
 * Da fuori si passa dal tunnel di Cloudflare, e il suo nome cambia quando il
 * tunnel riparte. Il telefono prova gli indirizzi che ha in tasca (vedi
 * [Indirizzi]); se sono tutti morti non ha modo di sapere quello nuovo — per
 * saperlo dovrebbe parlare col computer, e per parlarci gli serve l'indirizzo.
 * Quella sera l'app diceva «non riesco a parlare col computer» con il computer
 * acceso e raggiungibile, sotto un nome che lei non conosceva.
 *
 * ## Come funziona
 *
 * Il computer scrive i suoi indirizzi da fuori su una bacheca pubblica —
 * ntfy.sh, gratuita e senza account — sotto un nome ricavato dalla chiave di
 * questo telefono, e li firma con la stessa chiave. Qui, quando nessun
 * indirizzo risponde, si legge quel cartello, si controlla la firma, e si bussa
 * agli indirizzi nuovi con la solita `/io`.
 *
 * **Perche' ci si puo' fidare**: il nome non si indovina senza la chiave, la
 * firma non si falsifica senza la chiave, e comunque `/io` fa entrare solo il
 * computer che ci conosce.
 *
 * ⚠ **La stessa ricetta sta nel computer**, in `packages/gateway/src/cartello.ts`.
 * `CartelloTest` e `prova-cartello.mjs` calcolano lo stesso esempio e aspettano
 * gli stessi numeri: se una delle due cambia, cade la sua prova.
 */
object Cartello {

    /** La bacheca: la stessa del computer. */
    const val BACHECA = "https://ntfy.sh"

    private const val VERSIONE = 1

    /**
     * Non si chiede piu' di una volta al minuto per telefono. Il giro in
     * background passa di qui ogni volta che il computer non risponde, e un
     * computer spento per una notte non deve diventare cento domande alla
     * bacheca.
     */
    private const val NON_PIU_DI_OGNI_MS = 60_000L
    private val ultimaDomanda = mutableMapOf<String, Long>()

    private val rete = OkHttpClient.Builder()
        .connectTimeout(8, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .build()

    /** Il nome del cartello di questo telefono: si ricava dalla chiave, e non la rivela. */
    fun argomentoDi(token: String): String =
        "daprod_" + esadecimale(sha256("daprod-cartello-nome:$token")).take(32)

    /** La firma di un cartello: la stessa ricetta di `firmaDi` nel computer. */
    internal fun firmaDi(token: String, pc: String, quando: Long, basi: List<String>): String {
        val testo = (listOf("daprod-cartello/$VERSIONE", pc, quando.toString()) + basi)
            .joinToString("\n")
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(sha256("daprod-cartello-chiave:$token"), "HmacSHA256"))
        return esadecimale(mac.doFinal(testo.toByteArray(Charsets.UTF_8)))
    }

    /** Vero se quella firma e' davvero del computer che ha la nostra chiave. */
    internal fun firmaValida(
        token: String,
        pc: String,
        quando: Long,
        basi: List<String>,
        firma: String,
    ): Boolean = MessageDigest.isEqual(
        firmaDi(token, pc, quando, basi).toByteArray(),
        firma.lowercase().toByteArray(),
    )

    /**
     * ⚠ **Dal cartello si accettano solo indirizzi https.** Sono quelli da fuori:
     * il tunnel, o un nome fisso. Un indirizzo in chiaro scritto su una bacheca
     * pubblica non ha niente da fare in questo telefono, anche se firmato.
     */
    internal fun daTenere(basi: List<String>): List<String> =
        basi.map { it.trim().trimEnd('/') }.filter { it.startsWith("https://") }.distinct()

    /**
     * Gli indirizzi scritti sul cartello di questo telefono, dal piu' recente
     * cartello valido. Vuota se non c'e' niente, se la bacheca non risponde, o
     * se si e' chiesto meno di un minuto fa.
     */
    suspend fun leggi(token: String): List<String> = withContext(Dispatchers.IO) {
        val argomento = argomentoDi(token)
        val adesso = System.currentTimeMillis()
        synchronized(ultimaDomanda) {
            val prima = ultimaDomanda[argomento]
            if (prima != null && adesso - prima < NON_PIU_DI_OGNI_MS) return@withContext emptyList()
            ultimaDomanda[argomento] = adesso
        }
        try {
            val req = Request.Builder()
                .url("$BACHECA/$argomento/json?poll=1&since=24h")
                .build()
            rete.newCall(req).execute().use { res ->
                if (!res.isSuccessful) return@withContext emptyList()
                val righe = res.body?.string().orEmpty().lines()
                var migliore: Pair<Long, List<String>>? = null
                for (riga in righe) {
                    val letto = leggiRiga(riga, token) ?: continue
                    if (migliore == null || letto.first > migliore.first) migliore = letto
                }
                migliore?.second ?: emptyList()
            }
        } catch (_: Exception) {
            emptyList()
        }
    }

    /**
     * Una riga della bacheca: il cartello dentro, se c'e' e se la firma torna.
     * Visibile alle prove: `CartelloTest` le passa una riga vera, com'e' arrivata
     * da ntfy.sh l'11 settembre 2026.
     */
    internal fun leggiRiga(riga: String, token: String): Pair<Long, List<String>>? = try {
        val evento = JSONObject(riga)
        if (evento.optString("event") != "message") {
            null
        } else {
            val c = JSONObject(evento.optString("message"))
            val arr = c.optJSONArray("basi")
            val basi = (0 until (arr?.length() ?: 0)).mapNotNull { arr?.optString(it) }
            val pc = c.optString("pc")
            val quando = c.optLong("quando")
            if (c.optInt("v") == VERSIONE && pc.isNotBlank() &&
                firmaValida(token, pc, quando, basi, c.optString("firma"))
            ) {
                quando to daTenere(basi)
            } else {
                null
            }
        }
    } catch (_: Exception) {
        null
    }

    private fun sha256(testo: String): ByteArray =
        MessageDigest.getInstance("SHA-256").digest(testo.toByteArray(Charsets.UTF_8))

    private fun esadecimale(byte: ByteArray): String =
        byte.joinToString("") { "%02x".format(it) }
}
